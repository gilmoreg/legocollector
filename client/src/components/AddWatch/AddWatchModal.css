.Overlay {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0, 0, 0, 0.75);
}

.AddWatchModal {
  position: absolute;
  top: 100px;
  left: 50px;
  right: 50px;
  background-color: white;
  font-size: 40px;
  text-align: center;
  border-radius: 10px;
}

.AddWatchModal h3 {
  margin: 0;
  padding-bottom: 1rem;
}

.AddWatchModal input {
  width: 80%;
  margin-bottom: 1rem;
  text-align: center;
  border-radius: 5px;
  background-color: #CCCCCC;
}

.AddWatchModal small {
  font-size: 11px;
  color: #FF7F66;
  display: block;
}