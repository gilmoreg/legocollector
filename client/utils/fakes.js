/* eslint-disable import/prefer-default-export */
export const fakeWatch = {
  id: 0,
  image: 'test',
  url: 'test',
  title: 'test',
  added: Date.now(),
  stock_levels: [
    { datetime: Date.now(), id: 0, stock_level: 0 },
  ],
};
