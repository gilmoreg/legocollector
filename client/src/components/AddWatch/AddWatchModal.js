/* eslint-disable jsx-a11y/no-static-element-interactions */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ReactModal from 'react-modal';
import { connect } from 'react-redux';
import debounce from 'lodash.debounce';
import { API_URL } from '../../config';
import SearchResult from './SearchResult';
import { addWatch } from '../../state/actions';
import './AddWatchModal.css';

export class AddWatchModal extends Component {
  constructor(props) {
    super(props);
    this.submitForm = this.submitForm.bind(this);
    this.state = {
      searchTerm: '',
      searchResult: {},
      error: '',
    };
    this.search = debounce(query => this.queryAPI(query), 250);
    this.onInputChange = this.onInputChange.bind(this);
    this.queryAPI = this.queryAPI.bind(this);
    this.addWatch = this.addWatch.bind(this);
    this.displayError = this.displayError.bind(this);
  }

  onInputChange(event) {
    event.persist();
    if (event.target && event.target.value) {
      const query = event.target.value.trim();
      this.setState({ searchTerm: query });
      if (query) this.search(query);
    }
  }

  queryAPI(query) {
    fetch(`${API_URL}/legoset/search/${query}?token=${this.props.token}`)
      .then(res => res.json())
      .then((res) => {
        if (res && res.result) this.setState({ searchResult: res.result });
        if (res && res.error) this.displayError(res.error);
      })
      .catch(err => this.displayError(err));
  }

  submitForm(event) {
    event.preventDefault();
    this.search();
  }

  addWatch() {
    if (this.state.searchResult.id) {
      fetch(`${API_URL}/watches/add`, {
        method: 'POST',
        body: JSON.stringify({
          id: this.state.searchResult.id,
          token: this.props.token,
        }),
      })
        .then(res => res.json())
        .then(res => res.result)
        .then((watch) => {
          if (watch && watch.id) {
            this.props.dispatch(addWatch(watch));
            this.props.close();
          }
        })
        .catch(err => this.displayError(err));
    }
  }

  displayError(error) {
    this.setState({ error });
    setTimeout(() => {
      this.setState({ error: '' });
    }, 5000);
  }

  render() {
    return (
      <ReactModal
        isOpen={this.props.open}
        contentLabel="Add New Watch"
        overlayClassName="Overlay"
        className="AddWatchModal"
        shouldCloseOnOverlayClick
        onRequestClose={this.props.close}
        role="dialog"
        parentSelector={() => document.querySelector('#root')}
      >
        <form onSubmit={this.submitForm}>
          <h3>Lego ID:</h3>
          <input
            type="text"
            onChange={this.onInputChange}
          />
          { this.state.error ? <small>{this.state.error}</small> : ''}
        </form>
        {this.state.searchResult.id ?
          <SearchResult
            legoset={this.state.searchResult}
            onClick={this.addWatch}
          />
          : ''}
      </ReactModal>
    );
  }
}

AddWatchModal.defaultProps = {
  token: '',
  open: false,
  close: () => {},
  dispatch: () => {},
};

AddWatchModal.propTypes = {
  token: PropTypes.string,
  open: PropTypes.bool,
  close: PropTypes.func,
  dispatch: PropTypes.func,
};

const mapStateToProps = props => ({
  token: props.token,
});

export default connect(mapStateToProps)(AddWatchModal);
