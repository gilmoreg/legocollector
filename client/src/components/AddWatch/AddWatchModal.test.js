import React from 'react';
import fetchMock from 'fetch-mock';
import { shallow } from 'enzyme';
import { AddWatchModal } from './AddWatchModal';
import * as fakes from '../../../utils/fakes';

jest.useFakeTimers();

describe('AddWatchModal tests', () => {
  beforeEach(() => fetchMock.catch(500));
  afterEach(() => fetchMock.restore());
  
  it('renders without crashing', () => {
    const wrapper = shallow(
      <AddWatchModal open close={() => {}} />);
    expect(wrapper).toMatchSnapshot();
  });
  
  it('setSearchTerm sets searchTerm', () => {
    const wrapper = shallow(<AddWatchModal open close={() => {}} />);
    wrapper.instance().setState = jest.fn();
    const event = {
      target: { dataset: { id: '00000' }}
    }
    wrapper.instance().setSearchTerm(event);
    expect(wrapper.instance().setState).toHaveBeenCalled();
  });

  it('setSearchTerm errors on invalid input', () => {
    const wrapper = shallow(<AddWatchModal open close={() => {}} />);
    wrapper.instance().setState = jest.fn();
    // Empty object for event should throw an error
    wrapper.instance().setSearchTerm({});
    expect(wrapper.instance().setState).not.toHaveBeenCalled();
  });

  it('onInputChange should setState and search on valid input', () => {
    const wrapper = shallow(<AddWatchModal open close={() => {}} />);
    wrapper.instance().setState = jest.fn();
    wrapper.instance().search = jest.fn();
    const event = {
      persist: jest.fn(),
      target: { value: '00000' }
    }
    wrapper.instance().onInputChange(event);
    expect(wrapper.instance().setState).toHaveBeenCalled();
    expect(wrapper.instance().search).toHaveBeenCalled();
  });

  it('onInputChange should not setState or search on invalid input', () => {
    const wrapper = shallow(<AddWatchModal open close={() => {}} />);
    wrapper.instance().setState = jest.fn();
    wrapper.instance().search = jest.fn();
    const event = {
      persist: jest.fn(),
      target: { value: 'abc' } // characters fail digitTest
    }
    wrapper.instance().onInputChange(event);
    expect(wrapper.instance().setState).not.toHaveBeenCalled();
    expect(wrapper.instance().search).not.toHaveBeenCalled();
  });

  
  it('submitting form should query the API on a valid input', (done) => {
    const wrapper = shallow(<AddWatchModal open close={() => {}} />);
    wrapper.instance().queryAPI = jest.fn();
    // Set searchTerm to something that will pass digitTest
    wrapper.instance().setState({ searchTerm: '00000' });
    // Simulate a form submit
    wrapper.find('form').simulate('submit', { preventDefault () {} });
    // Expire timers in debounce
    jest.runAllTimers();
    expect(wrapper.instance().queryAPI).toHaveBeenCalled();
    done();    
  });

  it('queryAPI sets searchResult on good input', async (done) => {
    fetchMock.mock(/.legoset./, fakes.fakeSearchSuccess);
    const wrapper = shallow(<AddWatchModal open close={() => {}} />);
    // Set searchTerm to something that will pass digitTest
    wrapper.instance().setState({ searchTerm: '00000' });
    await wrapper.instance().queryAPI();
    expect(wrapper.instance().state.searchResult).toEqual(fakes.fakeSearchResult);
    done();
  });
  
  it('queryAPI displays error on invalid input', async (done) => {
    const wrapper = shallow(<AddWatchModal open close={() => {}} />);
    // Set searchTerm to something that will pass digitTest
    wrapper.instance().setState({ searchTerm: '0' });
    wrapper.instance().displayError = jest.fn();
    try {
      await wrapper.instance().queryAPI();
    } catch (e) {
      expect(wrapper.instance().displayError).toHaveBeenCalled();
      done();
    }
  });

  it('queryAPI displays error when API is down', async (done) => {
    fetchMock.mock(/.legoset./, 500);
    const wrapper = shallow(<AddWatchModal open close={() => {}} />);
    wrapper.instance().setState({ searchTerm: '00000' });
    wrapper.instance().displayError = jest.fn();
    await wrapper.instance().queryAPI();
    expect(wrapper.instance().displayError).toHaveBeenCalled();
    done();
  });

  it('queryAPI displays error when error returned from API', async (done) => {
    fetchMock.mock(/.legoset./, fakes.fakeSearchError);
    const wrapper = shallow(<AddWatchModal open close={() => {}} />);
    // Set searchTerm to something that will pass digitTest
    wrapper.instance().setState({ searchTerm: '00000' });
    wrapper.instance().displayError = jest.fn();
    await wrapper.instance().queryAPI();
    expect(wrapper.instance().displayError).toHaveBeenCalled();
    done();
  });

  it('addWatch should close modal if successful', async (done) => {
    fetchMock.mock(/.watches./, fakes.fakeAddWatchSuccess);
    const wrapper = shallow(<AddWatchModal open close={jest.fn()} />);
    wrapper.instance().setState({ searchResult: fakes.fakeSearchResult });
    await wrapper.instance().addWatch();
    expect(wrapper.instance().props.close).toHaveBeenCalled();
    done();
  });

  it('addWatch should displayError if unsuccessful', async (done) => {
    fetchMock.mock(/.watches./, fakes.fakeAddWatchError);
    const wrapper = shallow(<AddWatchModal open close={() => {}} />);
    wrapper.instance().setState({ searchResult: fakes.fakeSearchResult });
    wrapper.instance().displayError = jest.fn();
    await wrapper.instance().addWatch();
    expect(wrapper.instance().displayError).toHaveBeenCalled();
    done();
  });

  it('addWatch should displayError if API call fails', async (done) => {
    fetchMock.mock(/.watches./, 500);
    const wrapper = shallow(<AddWatchModal open close={() => {}} />);
    wrapper.instance().setState({ searchResult: fakes.fakeSearchResult });
    wrapper.instance().displayError = jest.fn();
    await wrapper.instance().addWatch();
    expect(wrapper.instance().displayError).toHaveBeenCalled();
    done();
  });
});

/*
Scratch:
setImmediate(() => {
  jest.runAllTimers();
  console.log(wrapper.update().instance().setState.mock);
  done();
});
// Expire timers in throttle
    // jest.runAllTimers();
    // expect(wrapper.instance().displayError).toHaveBeenCalled();
*/
