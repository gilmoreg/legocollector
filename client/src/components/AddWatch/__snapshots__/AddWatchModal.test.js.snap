// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddWatchModal
    close={[Function]}
    dispatch={[Function]}
    open={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "class",
    "props": Object {
      "ariaHideApp": true,
      "bodyOpenClassName": "ReactModal__Body--open",
      "children": <div
        className="AddWatchModalContent"
      >
        <form
          onSubmit={[Function]}
        >
          <header>
            Lego ID:
          </header>
          <Examples
            click={[Function]}
          />
          <input
            onChange={[Function]}
            placeholder={75105}
            type="text"
            value=""
          />
          
        </form>
        
      </div>,
      "className": "AddWatchModal",
      "closeTimeoutMS": 0,
      "contentLabel": "Add New Watch",
      "isOpen": true,
      "onRequestClose": [Function],
      "overlayClassName": "Overlay",
      "parentSelector": [Function],
      "portalClassName": "ReactModalPortal",
      "role": "dialog",
      "shouldCloseOnEsc": true,
      "shouldCloseOnOverlayClick": true,
      "shouldFocusAfterRender": true,
      "shouldReturnFocusAfterClose": true,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <form
            onSubmit={[Function]}
          >
            <header>
              Lego ID:
            </header>
            <Examples
              click={[Function]}
            />
            <input
              onChange={[Function]}
              placeholder={75105}
              type="text"
              value=""
            />
            
          </form>,
          "",
        ],
        "className": "AddWatchModalContent",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <header>
                Lego ID:
              </header>,
              <Examples
                click={[Function]}
              />,
              <input
                onChange={[Function]}
                placeholder={75105}
                type="text"
                value=""
              />,
              "",
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": "Lego ID:",
              },
              "ref": null,
              "rendered": "Lego ID:",
              "type": "header",
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "function",
              "props": Object {
                "click": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "onChange": [Function],
                "placeholder": 75105,
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            "",
          ],
          "type": "form",
        },
        "",
      ],
      "type": "div",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "class",
      "props": Object {
        "ariaHideApp": true,
        "bodyOpenClassName": "ReactModal__Body--open",
        "children": <div
          className="AddWatchModalContent"
        >
          <form
            onSubmit={[Function]}
          >
            <header>
              Lego ID:
            </header>
            <Examples
              click={[Function]}
            />
            <input
              onChange={[Function]}
              placeholder={75105}
              type="text"
              value=""
            />
            
          </form>
          
        </div>,
        "className": "AddWatchModal",
        "closeTimeoutMS": 0,
        "contentLabel": "Add New Watch",
        "isOpen": true,
        "onRequestClose": [Function],
        "overlayClassName": "Overlay",
        "parentSelector": [Function],
        "portalClassName": "ReactModalPortal",
        "role": "dialog",
        "shouldCloseOnEsc": true,
        "shouldCloseOnOverlayClick": true,
        "shouldFocusAfterRender": true,
        "shouldReturnFocusAfterClose": true,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <form
              onSubmit={[Function]}
            >
              <header>
                Lego ID:
              </header>
              <Examples
                click={[Function]}
              />
              <input
                onChange={[Function]}
                placeholder={75105}
                type="text"
                value=""
              />
              
            </form>,
            "",
          ],
          "className": "AddWatchModalContent",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <header>
                  Lego ID:
                </header>,
                <Examples
                  click={[Function]}
                />,
                <input
                  onChange={[Function]}
                  placeholder={75105}
                  type="text"
                  value=""
                />,
                "",
              ],
              "onSubmit": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": null,
                "nodeType": "host",
                "props": Object {
                  "children": "Lego ID:",
                },
                "ref": null,
                "rendered": "Lego ID:",
                "type": "header",
              },
              Object {
                "instance": null,
                "key": null,
                "nodeType": "function",
                "props": Object {
                  "click": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": null,
                "nodeType": "host",
                "props": Object {
                  "onChange": [Function],
                  "placeholder": 75105,
                  "type": "text",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              "",
            ],
            "type": "form",
          },
          "",
        ],
        "type": "div",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
